internal/routes/all_routes.go

```diff
@@ -23,15 +24,21 @@ func SetupRoutes(r *chi.Mux) {
 	r.Get("/ping", s.StatusService)
 	r.Route("/docs", func(r chi.Router) {
 		r.Use(redocMiddleware)
-		r.Get("/openapi.json", api.ServeOpenAPISpec)
+		r.Route("/openapi.json", func(r chi.Router) {
+			r.Use(middleware.ETagMiddleware(15*time.Minute, "json-spec", api.ETagJSONSpec))
```

Comment:
I've left the review here last time (sorry!)
I was trying to think of a way how we could introduce the middleware globally, but enable it per endpoint, but I guess this is the best way.

The only thing I've came up with, is to include the the tag in the return statement of the handler and middleware would be watching out for that, but it sounds bit ugly as we would change the Handler func definition just for that :thinking:

Reply:
So for now we are  caheching only in the openapi.json? and we are supposed to use this middleware in any endpoint which holds static information?

Reply:
Yeah this is the cleanest approach. Frankly, I don't see us using this for anything else than instance types and OpenAPI.

@adiabramovitch yes etag only works for content that does not change often (e.g. photos uploaded to a site). In our case instance types.

ETag is complementary to normal caching, it is not here to replace it at all. While it is possible to implement ETag with cache in tandem (could be timestamp of when some item was pulled from cache), I don't think it is worth the effort for our use case. If we cache something, let's just return it without ETag since it will be pretty fast anyway.

Reply:
Well it will be fast, but the trafic still needs to happen, right, so if we are caching it, it's probably good idea to etag it as well IMHO. 200KB traffic is IMHO quite some safe